src/main/scala/epic/parser/models/PositionalNeuralModel.scala

```diff
@@ -0,0 +1,501 @@
+                                             labelFeaturizer: RefinedFeaturizer[L, W, Feature],
+                                             val surfaceFeaturizer: Word2VecSurfaceFeaturizerIndexed[W],
+                                             depFeaturizer: Word2VecDepFeaturizerIndexed[W],
+                                             val layers: IndexedSeq[OutputTransform[Array[Int],DenseVector[Double]]#OutputLayer],
+                                             penultimateLayers: IndexedSeq[epic.dense.Transform.Layer[Array[Int],DenseVector[Double]]],
+                                             depLayers: IndexedSeq[OutputTransform[Array[Int],DenseVector[Double]]#OutputLayer],
+                                             val maybeSparseSurfaceFeaturizer: Option[IndexedSpanFeaturizer[L, L2, W]],
+                                             decoupledLayers: IndexedSeq[OutputTransform[Array[Int],DenseVector[Double]]#OutputLayer],
+                                             penultimateDecoupledLayers: IndexedSeq[epic.dense.Transform.Layer[Array[Int],DenseVector[Double]]],
+                                             val weights: DenseVector[Double],
+                                             val origPTModel: PositionalNeuralModel[L,L2,W]) extends Grammar[L, W] with Serializable {
+
+    val SpanLayerIdx = 0
+    val UnaryLayerIdx = 1
+    val BinaryLayerIdx = 2
+    val dcSpanFeatOffset = layers.map(_.index.size).foldLeft(0)(_ + _) + depLayers.map(_.index.size).foldLeft(0)(_ + _)
+    val dcUnaryFeatOffset = dcSpanFeatOffset + (if (decoupledLayers.size > 0) decoupledLayers(0).index.size else 0)
+    val dcBinaryFeatOffset = dcUnaryFeatOffset + (if (decoupledLayers.size > 0) decoupledLayers(1).index.size else 0)
+    
+    override def withPermissiveLexicon: Grammar[L, W] = {
+      new PositionalNeuralGrammar(topology, lexicon.morePermissive, refinedTopology, refinements, labelFeaturizer, surfaceFeaturizer,
+                                     depFeaturizer, layers, penultimateLayers, depLayers, maybeSparseSurfaceFeaturizer, decoupledLayers, penultimateDecoupledLayers, weights, origPTModel)
+    }
+
+
+    /**
+     * N.B. does not extracted expected counts for sparse features; this is done outside this loop
+     */
+    def extractEcounts(m: ParseMarginal[L, W], deriv: DenseVector[Double], scale: Double): Unit = {
+      val w = m.words
+      val length = w.length
+      val sspec = surfaceFeaturizer.anchor(w)
+      val depSpec = depFeaturizer.anchor(w)
+      val lspec = labelFeaturizer.anchor(w)
+      
+//      val maxTetraLen = ((w.size + 2) * (w.size + 3) * (w.size + 4))/6 + ((w.size + 1) * (w.size + 2))/2 + w.size + 2
+      
+      def tetra(begin: Int, split: Int, end: Int) = {
+        (end * (end + 1) * (end + 2))/6 + ((split + 1) * split / 2 + begin)
+      }
+      
+      // This representation appears to make things a bit faster?
```

Comment:
same strategy as packing a triangular array into a flat array. The nth tetrahedral number is (end * (end + 1) * (end + 2))/6. The nth triangular is ((split + 1) * split / 2, and you can pack \sum_{i<end} T_i triangular numbers before getting to the next one.

Inverting a tetrahedral number is really complicated, I think.